packages/@lwc/template-compiler/src/codegen/static-element-serializer.ts

```diff
@@ -39,25 +40,39 @@ function serializeAttrs(element: Element): string {
     const attrs: string[] = [];
     let hasClassAttr = false;
 
-    const collector = ({ name, value }: { name: string; value: string | boolean }) => {
+    const collector = ({
+        name,
+        value,
+        hasExpression,
+    }: {
+        name: string;
+        value: string | boolean;
+        hasExpression?: boolean;
+    }) => {
         let v = typeof value === 'string' ? templateStringEscape(value) : value;
 
         if (name === 'class') {
             hasClassAttr = true;
             v += '${0}';
         }
         if (typeof v === 'string') {
-            attrs.push(`${name}="${htmlEscape(v, true)}"`);
+            // Inject a placeholder where the staticPartId will go when an expression occurs.
+            // This is only needed for SSR to inject the expression value during serialization.
+            attrs.push(hasExpression ? `\${"${v}"}` : ` ${name}="${htmlEscape(v, true)}"`);
```

Comment:
When an expression is encountered, a token is inserted in its place, see the template compiler test fixtures for details.